File: spfx/packages/spfx-js/src/reader.ts
/**
 * SPIFReader — deserialize bytes into a SPIFDocument.
 */

import {
  MAGIC,
  SUPPORTED_VERSIONS,
  CHUNK_HEADER,
  CHUNK_PROVENANCE,
  CHUNK_SEMANTIC,
  CHUNK_TRACE,
  CHUNK_PAYLOAD,
  CHUNK_ALTS,
  CHUNK_DELTA,
  CHUNK_SIGNATURE,
  CHUNK_MULTISIG,
  CHUNK_CHECKSUM,
  FLAG_COMPRESSED,
  TAG_NODEREF,
  TRACE_POSTHOC,
} from './format.js';
import type {
  SPIFDocument,
  Node,
  TraceStep,
  Provenance,
  SemanticLayer,
  Alternative,
  Delta,
  Signature,
} from './types.js';
import {
  SPIFMagicError,
  SPIFVersionError,
  SPIFChecksumError,
  SPIFSignatureError,
  SPIFFormatError,
} from './errors.js';
import { sha256Sync, timingSafeEqual } from './checksum.js';
import { cborDecode, decodeNodeFromRaw, decodeStepFromRaw } from './cbor-utils.js';
import { Tag } from 'cbor-x';
import { createPublicKey, verify as verifySignatureRaw } from 'crypto';
import { inflateSync } from 'zlib';

// ---------------------------------------------------------------------------
// Chunk iteration
// ---------------------------------------------------------------------------

interface Chunk {
  type: number;
  data: Uint8Array;
  offset: number; // byte offset of the chunk header in the original buffer
}

function* iterChunks(data: Uint8Array, offset: number): Generator<Chunk> {
  const view = new DataView(data.buffer, data.byteOffset, data.byteLength);
  while (offset < data.length) {
    if (offset + 5 > data.length) {
      throw new SPIFFormatError('Truncated chunk header');
    }
    const chunkType = data[offset];
    const length = view.getUint32(offset + 1, false); // big-endian
    const payloadStart = offset + 5;
    const payloadEnd = payloadStart + length;
    if (payloadEnd > data.length) {
      throw new SPIFFormatError(
        `Chunk 0x${chunkType.toString(16).padStart(2, '0')} claims ${length} bytes but only ${data.length - payloadStart} remain`
      );
    }
    yield {
      type: chunkType,
      data: data.subarray(payloadStart, payloadEnd),
      offset,
    };
    offset = payloadEnd;
  }
}

// ---------------------------------------------------------------------------
// DAG validation
// ---------------------------------------------------------------------------

function validateTraceDAG(steps: TraceStep[]): void {
  const seen = new Set<string>();
  for (const s of steps) {
    if (seen.has(s.id)) {
      throw new SPIFFormatError(`Duplicate TraceStep id: '${s.id}'`);
    }
    seen.add(s.id);
  }

  const ids = new Set(steps.map(s => s.id));
  for (const s of steps) {
    for (const dep of s.deps) {
      if (!ids.has(dep)) {
        throw new SPIFFormatError(
          `TraceStep '${s.id}' references unknown dep '${dep}'`
        );
      }
    }
  }

  // Kahn's algorithm to detect cycles
  const inDegree = new Map<string, number>(steps.map(s => [s.id, 0]));
  const dependents = new Map<string, string[]>(steps.map(s => [s.id, []]));

  for (const s of steps) {
    for (const dep of s.deps) {
      inDegree.set(s.id, (inDegree.get(s.id) ?? 0) + 1);
      dependents.get(dep)!.push(s.id);
    }
  }

  const queue: string[] = [];
  for (const [id, deg] of inDegree) {
    if (deg === 0) queue.push(id);
  }

  let visited = 0;
  while (queue.length > 0) {
    const node = queue.shift()!;
    visited++;
    for (const child of dependents.get(node) ?? []) {
      const deg = (inDegree.get(child) ?? 0) - 1;
      inDegree.set(child, deg);
      if (deg === 0) queue.push(child);
    }
  }

  if (visited !== steps.length) {
    const cycleMembers = [...inDegree.entries()]
      .filter(([, deg]) => deg > 0)
      .map(([id]) => id);
    throw new SPIFFormatError(
      `Cycle detected in trace DAG involving steps: ${cycleMembers.join(', ')}`
    );
  }
}

// ---------------------------------------------------------------------------
// Safe CBOR loader
// ---------------------------------------------------------------------------

function cborLoad(data: Uint8Array, chunkName: string, compressed = false): unknown {
  try {
    const raw = compressed ? new Uint8Array(inflateSync(data)) : data;
    return cborDecode(raw);
  } catch (e) {
    throw new SPIFFormatError(
      `Malformed CBOR in ${chunkName} chunk: ${e instanceof Error ? e.message : String(e)}`
    );
  }
}

function findFirstAuthChunkOffset(data: Uint8Array): number | null {
  let offset = MAGIC.length + 2;
  const view = new DataView(data.buffer, data.byteOffset, data.byteLength);

  while (offset < data.length) {
    if (offset + 5 > data.length) return null;
    const chunkType = data[offset];
    const length = view.getUint32(offset + 1, false);
    if (chunkType === CHUNK_SIGNATURE || chunkType === CHUNK_MULTISIG) {
      return offset;
    }
    if (chunkType === CHUNK_CHECKSUM) {
      return null;
    }
    offset += 5 + length;
  }

  return null;
}

const ED25519_SPKI_PREFIX = Buffer.from('302a300506032b6570032100', 'hex');

// ---------------------------------------------------------------------------
// Main reader
// ---------------------------------------------------------------------------

export class SPIFReader {
  private requireSignature: boolean;

  constructor(options: { requireSignature?: boolean } = {}) {
    this.requireSignature = options.requireSignature ?? false;
  }

  decode(data: Uint8Array): SPIFDocument {
    // 1. Validate magic
    if (data.length < MAGIC.length + 2) {
      throw new SPIFMagicError('File too short to be a SPIF document');
    }
    for (let i = 0; i < MAGIC.length; i++) {
      if (data[i] !== MAGIC[i]) {
        throw new SPIFMagicError(
          `Bad magic bytes: ${Buffer.from(data.subarray(0, MAGIC.length)).toString('hex')} (expected ${Buffer.from(MAGIC).toString('hex')})`
        );
      }
    }

    const version = data[MAGIC.length];
    if (!SUPPORTED_VERSIONS.has(version)) {
      throw new SPIFVersionError(`Unsupported SPIF version: ${version}`);
    }

    // 2. Find CHECKSUM chunk
    let checksumOffset: number | null = null;
    let scanOffset = MAGIC.length + 2;
    const scanView = new DataView(data.buffer, data.byteOffset, data.byteLength);

    while (scanOffset < data.length) {
      if (scanOffset + 5 > data.length) break;
      const chunkType = data[scanOffset];
      const length = scanView.getUint32(scanOffset + 1, false);
      if (chunkType === CHUNK_CHECKSUM) {
        checksumOffset = scanOffset;
        break;
      }
      scanOffset += 5 + length;
    }

    if (checksumOffset === null) {
      throw new SPIFFormatError('No CHECKSUM chunk found');
    }

    // 3. Validate checksum
    const body = data.subarray(0, checksumOffset);
    const csumLen = scanView.getUint32(checksumOffset + 1, false);
    const storedChecksum = data.subarray(checksumOffset + 5, checksumOffset + 5 + csumLen);
    const computed = sha256Sync(body);

    if (!timingSafeEqual(storedChecksum, computed)) {
      throw new SPIFChecksumError(
        `Checksum mismatch: stored=${Buffer.from(storedChecksum).toString('hex')}, computed=${Buffer.from(computed).toString('hex')}`
      );
    }

    // 4. Parse all chunks up to CHECKSUM
    const chunks = new Map<number, Uint8Array[]>();
    for (const chunk of iterChunks(data, MAGIC.length + 2)) {
      if (chunk.type === CHUNK_CHECKSUM) break;
      if (!chunks.has(chunk.type)) chunks.set(chunk.type, []);
      chunks.get(chunk.type)!.push(chunk.data);
    }

    let compressed = false;
    if (chunks.has(CHUNK_HEADER)) {
      const header = cborLoad(chunks.get(CHUNK_HEADER)![0], 'HEADER') as Record<string, unknown>;
      if (typeof header !== 'object' || header === null || Array.isArray(header)) {
        throw new SPIFFormatError('HEADER chunk must decode to a dict');
      }
      compressed = (((header['flags2'] as number) ?? 0) & FLAG_COMPRESSED) !== 0;
    }

    // 5. PAYLOAD (required)
    if (!chunks.has(CHUNK_PAYLOAD)) {
      throw new SPIFFormatError('Missing required PAYLOAD chunk');
    }

    const rawPayload = cborLoad(chunks.get(CHUNK_PAYLOAD)![0], 'PAYLOAD', compressed);
    if (!Array.isArray(rawPayload)) {
      throw new SPIFFormatError('PAYLOAD chunk must decode to a list');
    }
    const payload: Node[] = rawPayload.map((n: unknown) =>
      decodeNodeFromRaw(n as Record<string, unknown>)
    );

    if (payload.length === 0) {
      throw new SPIFFormatError('PAYLOAD must contain at least one node');
    }

    // 6. PROVENANCE
    let provenance: Provenance | undefined;
    if (chunks.has(CHUNK_PROVENANCE)) {
      const p = cborLoad(chunks.get(CHUNK_PROVENANCE)![0], 'PROVENANCE') as Record<string, unknown>;
      if (typeof p !== 'object' || p === null || Array.isArray(p)) {
        throw new SPIFFormatError('PROVENANCE chunk must decode to a dict');
      }
      provenance = {
        source_model: (p['source_model'] as string) ?? '',
        model_version: (p['model_version'] as string) ?? '',
        temperature: (p['temperature'] as number) ?? 0.0,
        input_hash: (p['input_hash'] as string) ?? '',
        context_ref: (p['context_ref'] as string) ?? '',
        timestamp_ms: (p['timestamp_ms'] as number) ?? 0,
      };
    }

    // 7. SEMANTIC
    let semantic: SemanticLayer | undefined;
    if (chunks.has(CHUNK_SEMANTIC)) {
      const s = cborLoad(chunks.get(CHUNK_SEMANTIC)![0], 'SEMANTIC', compressed) as Record<string, unknown>;
      if (typeof s !== 'object' || s === null || Array.isArray(s)) {
        throw new SPIFFormatError('SEMANTIC chunk must decode to a dict');
      }
      const emb = (s['embedding'] as number[]) ?? [];
      semantic = {
        embedding: emb,
        embedding_model: (s['embedding_model'] as string) ?? '',
        covariance: (s['covariance'] as number[][]) ?? [],
      };
    }

    // 8. TRACE (v0.2: wrapped dict; v0.1: bare list)
    let trace: TraceStep[] = [];
    let traceMethod = TRACE_POSTHOC;
    if (chunks.has(CHUNK_TRACE)) {
      const raw = cborLoad(chunks.get(CHUNK_TRACE)![0], 'TRACE', compressed);
      if (Array.isArray(raw)) {
        // v0.1 format: bare list of steps
        trace = raw.map((st: unknown) => decodeStepFromRaw(st as Record<string, unknown>));
      } else {
        // v0.2 format: {method, steps}
        const rawObj = raw as Record<string, unknown>;
        traceMethod = (rawObj['method'] as string) ?? TRACE_POSTHOC;
        trace = ((rawObj['steps'] as unknown[]) ?? []).map((st: unknown) =>
          decodeStepFromRaw(st as Record<string, unknown>)
        );
      }
    }

    // DAG validation
    if (trace.length > 0) {
      validateTraceDAG(trace);
    }

    // 9. ALTS (v0.2: wrapped dict; v0.1: bare list)
    const alternatives: Alternative[] = [];
    if (chunks.has(CHUNK_ALTS)) {
      const raw = cborLoad(chunks.get(CHUNK_ALTS)![0], 'ALTS', compressed);
      if (Array.isArray(raw)) {
        // v0.1 format
        for (const a of raw as Array<Record<string, unknown>>) {
          alternatives.push({
            weight: a['weight'] as number,
            nodes: (a['nodes'] as Array<Record<string, unknown>>).map(decodeNodeFromRaw),
            normalized: true,
          });
        }
      } else {
        // v0.2 format: {normalized, alts}
        const rawObj = raw as Record<string, unknown>;
        const normalized = (rawObj['normalized'] as boolean) ?? true;
        for (const a of (rawObj['alts'] as Array<Record<string, unknown>>) ?? []) {
          alternatives.push({
            weight: a['weight'] as number,
            nodes: (a['nodes'] as Array<Record<string, unknown>>).map(decodeNodeFromRaw),
            normalized,
          });
        }
        // Validate weights if normalized
        if (normalized && alternatives.length > 0) {
          const total = alternatives.reduce((sum, a) => sum + a.weight, 0);
          if (Math.abs(total - 1.0) > 0.01) {
            throw new SPIFFormatError(
              `ALTS normalized=True but weights sum to ${total.toFixed(4)}, expected 1.0 ± 0.01`
            );
          }
        }
      }
    }

    // 10. DELTA
    let delta: Delta | undefined;
    if (chunks.has(CHUNK_DELTA)) {
      const d = cborLoad(chunks.get(CHUNK_DELTA)![0], 'DELTA', compressed) as Record<string, unknown>;
      if (typeof d !== 'object' || d === null || Array.isArray(d)) {
        throw new SPIFFormatError('DELTA chunk must decode to a dict');
      }
      delta = {
        base_hash: d['base_hash'] as string,
        changes: d['changes'],
      };
    }

    // 11. SIGNATURE (v0.2)
    let signature: Signature | undefined;
    if (chunks.has(CHUNK_SIGNATURE)) {
      const s = cborLoad(chunks.get(CHUNK_SIGNATURE)![0], 'SIGNATURE') as Record<string, unknown>;
      if (typeof s !== 'object' || s === null || Array.isArray(s)) {
        throw new SPIFFormatError('SIGNATURE chunk must decode to a dict');
      }
      let sigBytes = s['signature'];
      if (sigBytes instanceof Tag) sigBytes = sigBytes.value;
      if (!(sigBytes instanceof Uint8Array)) {
        sigBytes = new Uint8Array(sigBytes as ArrayBuffer);
      }
      signature = {
        algorithm: (s['algorithm'] as string) ?? 'ed25519',
        signer: (s['signer'] as string) ?? '',
        signature: sigBytes as Uint8Array,
        key_id: (s['key_id'] as string) ?? '',
      };
    }

    // 12. MULTISIG (v0.2+)
    const signatures: Signature[] = [];
    if (chunks.has(CHUNK_MULTISIG)) {
      const raw = cborLoad(chunks.get(CHUNK_MULTISIG)![0], 'MULTISIG');
      if (!Array.isArray(raw)) {
        throw new SPIFFormatError('MULTISIG chunk must decode to a list');
      }
      for (const s of raw as Array<Record<string, unknown>>) {
        let sigBytes = s['signature'];
        if (sigBytes instanceof Tag) sigBytes = sigBytes.value;
        if (!(sigBytes instanceof Uint8Array)) {
          sigBytes = new Uint8Array(sigBytes as ArrayBuffer);
        }
        signatures.push({
          algorithm: (s['algorithm'] as string) ?? 'ed25519',
          signer: (s['signer'] as string) ?? '',
          signature: sigBytes as Uint8Array,
          key_id: (s['key_id'] as string) ?? '',
        });
      }
    }

    const doc: SPIFDocument = {
      payload,
      provenance,
      semantic,
      trace,
      trace_method: traceMethod,
      alternatives,
      delta,
      signature,
      signatures,
    };

    if (this.requireSignature) {
      if (!doc.signature && doc.signatures.length === 0) {
        throw new SPIFSignatureError(
          'Document is unsigned. SPIFReader was created with requireSignature=true'
        );
      }
      this.verifySignaturesInternal(data, doc);
    }

    return doc;
  }

  private verifySignaturesInternal(data: Uint8Array, doc: SPIFDocument): boolean {
    const signatures: Signature[] = [];
    if (doc.signature) signatures.push(doc.signature);
    signatures.push(...doc.signatures);

    if (signatures.length === 0) {
      return false;
    }

    const authOffset = findFirstAuthChunkOffset(data);
    if (authOffset === null) {
      throw new SPIFSignatureError(
        'Document advertises signature data but no SIGNATURE/MULTISIG chunk was found before CHECKSUM'
      );
    }
    const bodyToSign = Buffer.from(data.subarray(0, authOffset));

    for (const sig of signatures) {
      if (sig.algorithm !== 'ed25519') {
        throw new SPIFSignatureError(
          `Unsupported signature algorithm '${sig.algorithm}'; only ed25519 is supported`
        );
      }

      try {
        const rawPublicKey = Buffer.from(sig.signer, 'base64');
        if (rawPublicKey.length !== 32) {
          throw new Error('signer public key must decode to 32 bytes');
        }
        const publicKey = createPublicKey({
          key: Buffer.concat([ED25519_SPKI_PREFIX, rawPublicKey]),
          format: 'der',
          type: 'spki',
        });
        const ok = verifySignatureRaw(null, bodyToSign, publicKey, Buffer.from(sig.signature));
        if (!ok) {
          throw new Error('invalid signature');
        }
      } catch (e) {
        throw new SPIFSignatureError(
          `Signature verification failed: ${e instanceof Error ? e.message : String(e)}`
        );
      }
    }

    return true;
  }

  verifySignature(data: Uint8Array): boolean {
    const doc = this.decode(data);
    return this.verifySignaturesInternal(data, doc);
  }
}

File: spif-rust/src/reader.rs
use crate::types::*;
use anyhow::{anyhow, Result};
use base64::{engine::general_purpose::STANDARD, Engine as _};
use byteorder::{BigEndian, ReadBytesExt};
use ciborium::de::from_reader;
use ed25519_dalek::{Signature as Ed25519Signature, Verifier, VerifyingKey};
use flate2::read::ZlibDecoder;
use sha2::{Digest, Sha256};
use std::borrow::Cow;
use std::collections::{HashMap, HashSet, VecDeque};
use std::io::{Cursor, Read};

/// Deserializes SPIF documents from raw bytes.
///
/// # Security model
///
/// SHA-256 checksums detect *accidental* corruption (bit rot, truncation).
/// They do **not** protect against an attacker who modifies content and
/// recomputes the checksum.  Set `require_signature = true` in production
/// pipelines where tamper-evidence against intentional substitution matters.
pub struct SPIFReader {
    pub require_signature: bool,
}

impl Default for SPIFReader {
    fn default() -> Self {
        Self {
            require_signature: false,
        }
    }
}

impl SPIFReader {
    pub fn new() -> Self {
        Self::default()
    }

    /// Create a reader that rejects unsigned documents.
    pub fn strict() -> Self {
        Self {
            require_signature: true,
        }
    }

    pub fn read(&self, data: &[u8]) -> Result<SPIFDocument> {
        // 1. Minimum length: magic (8) + version (1) + flags (1)
        if data.len() < MAGIC.len() + 2 {
            return Err(anyhow!("Data too short to be a SPIF document"));
        }

        // 2. Magic bytes
        if &data[0..MAGIC.len()] != MAGIC {
            return Err(anyhow!(
                "Invalid magic bytes: {} (expected {})",
                hex::encode(&data[0..MAGIC.len()]),
                hex::encode(MAGIC)
            ));
        }

        let mut cursor = Cursor::new(&data[MAGIC.len()..]);
        let version = cursor.read_u8()?;
        let _flags = cursor.read_u8()?;

        // 3. Version check
        if !SUPPORTED_VERSIONS.contains(&version) {
            return Err(anyhow!("Unsupported SIF version: 0x{:02x}", version));
        }

        let mut document_data = SPIFDocument {
            payload: Vec::new(),
            provenance: None,
            semantic: None,
            trace: Vec::new(),
            trace_method: "post-hoc".to_string(),
            alternatives: Vec::new(),
            delta: None,
            signature: None,
            signatures: Vec::new(),
            task_info: None,
        };

        let mut checksum_seen = false;
        let mut compressed = false; // set when FLAG_COMPRESSED is seen in HEADER chunk
        let mut current_pos = MAGIC.len() + 2;

        while current_pos < data.len() {
            let mut chunk_cursor = Cursor::new(&data[current_pos..]);
            let chunk_type = chunk_cursor.read_u8()?;
            let chunk_len = chunk_cursor.read_u32::<BigEndian>()?;
            let payload_start = current_pos + 5;
            let payload_end = payload_start + chunk_len as usize;

            if payload_end > data.len() {
                return Err(anyhow!(
                    "Chunk 0x{:02x} claims {} bytes but only {} remain",
                    chunk_type,
                    chunk_len,
                    data.len().saturating_sub(payload_start)
                ));
            }

            let payload = &data[payload_start..payload_end];

            match chunk_type {
                CHUNK_HEADER => {
                    // Parse flags2 to detect compression.
                    // HEADER CBOR is: {"version": u8, "flags": u8, "flags2": u8, "created_ms": u64}
                    #[derive(serde::Deserialize, Default)]
                    struct HeaderPayload {
                        #[serde(default)]
                        flags2: u8,
                    }
                    if let Ok(h) = from_reader::<HeaderPayload, _>(payload) {
                        if h.flags2 & FLAG_COMPRESSED != 0 {
                            compressed = true;
                        }
                    }
                }
                CHUNK_PROVENANCE => {
                    // PROVENANCE is never compressed — read raw.
                    document_data.provenance = Some(from_reader(payload)?);
                }
                CHUNK_SEMANTIC => {
                    document_data.semantic =
                        Some(from_reader(&*cbor_payload(payload, compressed)?)?);
                }
                CHUNK_TRACE => {
                    #[derive(serde::Deserialize)]
                    struct TraceWrapper {
                        method: String,
                        steps: Vec<TraceStep>,
                    }
                    let tw: TraceWrapper = from_reader(&*cbor_payload(payload, compressed)?)?;
                    document_data.trace_method = tw.method;
                    document_data.trace = tw.steps;
                }
                CHUNK_PAYLOAD => {
                    document_data.payload = from_reader(&*cbor_payload(payload, compressed)?)?;
                }
                CHUNK_ALTS => {
                    // The outer wrapper carries `normalized` (Python format).
                    // Each alt only has `weight` + `nodes` in Python-written files.
                    // We read the outer normalized and apply it to each alternative.
                    #[derive(serde::Deserialize)]
                    #[allow(dead_code)] // `normalized` is deserialized but outer wrapper takes precedence
                    struct AltItem {
                        weight: f64,
                        nodes: Vec<Node>,
                        // Per-element normalized (present in Rust-written files); overridden by
                        // the outer wrapper's normalized to match Python format semantics.
                        #[serde(default = "crate::types::default_true")]
                        normalized: bool,
                    }
                    #[derive(serde::Deserialize)]
                    struct AltsWrapper {
                        #[serde(default = "crate::types::default_true")]
                        normalized: bool,
                        alts: Vec<AltItem>,
                    }
                    let aw: AltsWrapper = from_reader(&*cbor_payload(payload, compressed)?)?;
                    let outer_normalized = aw.normalized;
                    document_data.alternatives = aw
                        .alts
                        .into_iter()
                        .map(|a| Alternative {
                            weight: a.weight,
                            nodes: a.nodes,
                            normalized: outer_normalized,
                        })
                        .collect();
                }
                CHUNK_DELTA => {
                    document_data.delta = Some(from_reader(&*cbor_payload(payload, compressed)?)?);
                }
                CHUNK_SIGNATURE => {
                    document_data.signature = Some(from_reader(payload)?);
                }
                CHUNK_MULTISIG => {
                    document_data.signatures = from_reader(payload)?;
                }
                CHUNK_TASK => {
                    document_data.task_info = Some(from_reader(payload)?);
                }
                CHUNK_CHECKSUM => {
                    let expected_checksum = payload;
                    let body = &data[0..current_pos];
                    let mut hasher = Sha256::new();
                    hasher.update(body);
                    let actual_checksum = hasher.finalize();
                    if actual_checksum.as_slice() != expected_checksum {
                        return Err(anyhow!(
                            "Checksum mismatch: stored={}, computed={}",
                            hex::encode(expected_checksum),
                            hex::encode(actual_checksum)
                        ));
                    }
                    checksum_seen = true;
                }
                _ => {
                    // Unknown chunk — skip for forward compatibility.
                }
            }

            current_pos = payload_end;
        }

        // 4. CHECKSUM is mandatory
        if !checksum_seen {
            return Err(anyhow!(
                "No CHECKSUM chunk found — document may be truncated or corrupted"
            ));
        }

        // 5. PAYLOAD is mandatory (spec requires at least one node)
        if document_data.payload.is_empty() {
            return Err(anyhow!("Missing or empty required PAYLOAD chunk"));
        }

        // 6. Validate Distribution fields in all nodes and trace steps
        for node in &document_data.payload {
            node.confidence
                .validate()
                .map_err(|e| anyhow!("Node '{}': {}", node.id, e))?;
        }
        for step in &document_data.trace {
            step.confidence
                .validate()
                .map_err(|e| anyhow!("TraceStep '{}': {}", step.id, e))?;
        }

        // 7. DAG cycle/dangling-dep validation
        validate_payload_dag(&document_data.payload)?;
        if !document_data.trace.is_empty() {
            validate_trace_dag(&document_data.trace)?;
        }

        // 8. Signature enforcement
        let mut signatures: Vec<&crate::types::Signature> = Vec::new();
        if let Some(signature) = document_data.signature.as_ref() {
            signatures.push(signature);
        }
        signatures.extend(document_data.signatures.iter());

        if self.require_signature {
            if signatures.is_empty() {
                return Err(anyhow!(
                    "Document is unsigned. SPIFReader was created with require_signature=true \
                     — sign the document before distributing it."
                ));
            }
            self.verify_signatures_internal(&signatures, data)?;
        }

        Ok(document_data)
    }

    fn verify_signatures_internal(&self, signatures: &[&crate::types::Signature], data: &[u8]) -> Result<()> {
        if signatures.is_empty() {
            return Ok(());
        }

        let auth_offset = find_first_auth_chunk_offset(data).ok_or_else(|| {
            anyhow!(
                "Document advertises signature data but no SIGNATURE/MULTISIG chunk \
                 was found before CHECKSUM"
            )
        })?;
        let body_to_sign = &data[..auth_offset];

        for signature in signatures {
            if signature.algorithm != "ed25519" {
                return Err(anyhow!(
                    "Unsupported signature algorithm '{}'; only ed25519 is supported",
                    signature.algorithm
                ));
            }

            let pub_bytes = STANDARD.decode(signature.signer.as_bytes()).map_err(|e| {
                anyhow!("Signature verification failed: invalid base64 signer: {e}")
            })?;
            let pub_bytes: [u8; 32] = pub_bytes.as_slice().try_into().map_err(|_| {
                anyhow!("Signature verification failed: signer public key must decode to 32 bytes")
            })?;
            let verifying_key = VerifyingKey::from_bytes(&pub_bytes)
                .map_err(|e| anyhow!("Signature verification failed: invalid public key: {e}"))?;
            let signature_bytes = Ed25519Signature::try_from(signature.signature.as_ref())
                .map_err(|e| {
                    anyhow!("Signature verification failed: invalid signature bytes: {e}")
                })?;
            verifying_key
                .verify(body_to_sign, &signature_bytes)
                .map_err(|e| anyhow!("Signature verification failed: invalid signature: {e}"))?;
        }

        Ok(())
    }

    /// Verify all present ed25519 signatures against the raw signing body.
    ///
    /// The signing body is every byte from MAGIC through the byte immediately
    /// before the first SIGNATURE or MULTISIG chunk. Returns `Ok(false)` when
    /// the document is unsigned, and an error if any present signature fails.
    pub fn verify_signature(&self, data: &[u8]) -> Result<bool> {
        let document = self.read(data)?;

        let mut signatures: Vec<&crate::types::Signature> = Vec::new();
        if let Some(signature) = document.signature.as_ref() {
            signatures.push(signature);
        }
        signatures.extend(document.signatures.iter());

        if signatures.is_empty() {
            return Ok(false);
        }

        self.verify_signatures_internal(&signatures, data)?;
        Ok(true)
    }
}

/// Return the CBOR payload bytes for a chunk, decompressing with zlib if needed.
/// PROVENANCE, SIGNATURE, MULTISIG, and CHECKSUM are never compressed; callers
/// that handle those pass the raw slice directly rather than calling this.
///
/// Returns `Cow::Borrowed` when uncompressed (zero copy) and `Cow::Owned` after
/// decompression. All callers immediately call `.as_slice()` and pass to
/// `from_reader()`, so the lifetime is always satisfied.
fn cbor_payload<'a>(payload: &'a [u8], compressed: bool) -> Result<Cow<'a, [u8]>> {
    if !compressed {
        return Ok(Cow::Borrowed(payload));
    }
    let mut decoder = ZlibDecoder::new(payload);
    let mut decompressed = Vec::new();
    decoder
        .read_to_end(&mut decompressed)
        .map_err(|e| anyhow!("zlib decompress failed: {}", e))?;
    Ok(Cow::Owned(decompressed))
}

fn find_first_auth_chunk_offset(data: &[u8]) -> Option<usize> {
    let mut pos = MAGIC.len() + 2;
    while pos + 5 <= data.len() {
        let chunk_type = data[pos];
        let len = u32::from_be_bytes(data[pos + 1..pos + 5].try_into().ok()?) as usize;
        if chunk_type == CHUNK_SIGNATURE || chunk_type == CHUNK_MULTISIG {
            return Some(pos);
        }
        if chunk_type == CHUNK_CHECKSUM {
            return None;
        }
        pos = pos.checked_add(5 + len)?;
    }
    None
}

/// Validates the trace DAG using Kahn's topological sort algorithm.
/// Rejects duplicate step IDs, dangling dependency references, and cycles.
fn validate_trace_dag(steps: &[TraceStep]) -> Result<()> {
    // Check for duplicate IDs
    let mut seen: HashSet<&str> = HashSet::new();
    for s in steps {
        if !seen.insert(s.id.as_str()) {
            return Err(anyhow!("Duplicate TraceStep id: '{}'", s.id));
        }
    }

    // Fast-path: if no steps have any dependencies, return early
    if !steps.iter().any(|s| !s.deps.is_empty()) {
        return Ok(());
    }

    // Check for dangling deps
    let ids: HashSet<&str> = steps.iter().map(|s| s.id.as_str()).collect();
    for s in steps {
        for dep in &s.deps {
            if !ids.contains(dep.as_str()) {
                return Err(anyhow!(
                    "TraceStep '{}' references unknown dep '{}'",
                    s.id,
                    dep
                ));
            }
        }
    }

    // Kahn's algorithm — cycle detection
    let mut in_degree: HashMap<&str, usize> =
        steps.iter().map(|s| (s.id.as_str(), 0usize)).collect();
    let mut dependents: HashMap<&str, Vec<&str>> =
        steps.iter().map(|s| (s.id.as_str(), vec![])).collect();

    for s in steps {
        for dep in &s.deps {
            *in_degree.get_mut(s.id.as_str()).unwrap() += 1;
            dependents
                .get_mut(dep.as_str())
                .unwrap()
                .push(s.id.as_str());
        }
    }

    let mut queue: VecDeque<&str> = in_degree
        .iter()
        .filter(|(_, &deg)| deg == 0)
        .map(|(&id, _)| id)
        .collect();

    let mut visited = 0usize;
    while let Some(node) = queue.pop_front() {
        visited += 1;
        for &child in dependents.get(node).map(Vec::as_slice).unwrap_or(&[]) {
            let deg = in_degree.get_mut(child).unwrap();
            *deg -= 1;
            if *deg == 0 {
                queue.push_back(child);
            }
        }
    }

    if visited != steps.len() {
        let cycle_members: Vec<&str> = in_degree
            .iter()
            .filter(|(_, &deg)| deg > 0)
            .map(|(&id, _)| id)
            .collect();
        return Err(anyhow!(
            "Cycle detected in trace DAG involving steps: {:?}",
            cycle_members
        ));
    }

    Ok(())
}

/// Validates the payload DAG references using Kahn's topological sort algorithm.
/// Rejects duplicate node IDs, dangling references, and cycles.
fn validate_payload_dag(nodes: &[Node]) -> Result<()> {
    // Check for duplicate IDs
    let mut seen: HashSet<&str> = HashSet::new();
    for n in nodes {
        if !seen.insert(n.id.as_str()) {
            return Err(anyhow!("Duplicate Node id: '{}'", n.id));
        }
    }

    // Fast-path: if no nodes have references, return early
    if !nodes.iter().any(|n| !n.refs.is_empty()) {
        return Ok(());
    }

    // Check for dangling refs
    let ids: HashSet<&str> = nodes.iter().map(|n| n.id.as_str()).collect();
    for n in nodes {
        for ref_id in &n.refs {
            if !ids.contains(ref_id.as_str()) {
                return Err(anyhow!(
                    "Node '{}' references unknown node_id '{}'",
                    n.id,
                    ref_id
                ));
            }
        }
    }

    // Kahn's algorithm — cycle detection
    let mut in_degree: HashMap<&str, usize> =
        nodes.iter().map(|n| (n.id.as_str(), 0usize)).collect();
    let mut dependents: HashMap<&str, Vec<&str>> =
        nodes.iter().map(|n| (n.id.as_str(), vec![])).collect();

    for n in nodes {
        for ref_id in &n.refs {
            *in_degree.get_mut(n.id.as_str()).unwrap() += 1;
            dependents
                .get_mut(ref_id.as_str())
                .unwrap()
                .push(n.id.as_str());
        }
    }

    let mut queue: VecDeque<&str> = in_degree
        .iter()
        .filter(|(_, &deg)| deg == 0)
        .map(|(&id, _)| id)
        .collect();

    let mut visited = 0usize;
    while let Some(node) = queue.pop_front() {
        visited += 1;
        for &child in dependents.get(node).map(Vec::as_slice).unwrap_or(&[]) {
            let deg = in_degree.get_mut(child).unwrap();
            *deg -= 1;
            if *deg == 0 {
                queue.push_back(child);
            }
        }
    }

    if visited != nodes.len() {
        let cycle_members: Vec<&str> = in_degree
            .iter()
            .filter(|(_, &deg)| deg > 0)
            .map(|(&id, _)| id)
            .collect();
        return Err(anyhow!(
            "Cycle detected in payload Node refs involving nodes: {:?}",
            cycle_members
        ));
    }

    Ok(())
}
